### PR TITLE
Add signal handler to reap zombie process.

### PR DESCRIPTION
**Pull request by chen3933**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit [53be21b](https://github.com/aws/deep-learning-containers/pull/4249/commits/53be21b5803dfdbf8efeb55d9e8cb969785ed840) for my changes.

<details>
<summary>Confused on how to run tests? Try using the helper utility...</summary>

Assuming your remote is called `origin` (you can find out more with `git remote -v`)...
  
- Run default builds and tests for a particular buildspec - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -cp origin`

- Enable specific tests for a buildspec or set of buildspecs - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -t sanity_tests -cp origin`

- Restore TOML file when ready to merge

`python src/prepare_dlc_dev_environment.py -rcp origin`
</details>

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [x] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Fill out the template and click the checkbox of the builds you'd like to execute

*Note: Replace with <X.Y> with the major.minor framework version (i.e. 2.2) you would like to start.*

- [ ] build_pytorch_training_<X.Y>_sm
- [ ] build_pytorch_training_<X.Y>_ec2

- [ ] build_pytorch_inference_<X.Y>_sm
- [ ] build_pytorch_inference_<X.Y>_ec2
- [ ] build_pytorch_inference_<X.Y>_graviton

- [ ] build_tensorflow_training_<X.Y>_sm
- [ ] build_tensorflow_training_<X.Y>_ec2

- [ ] build_tensorflow_inference_<X.Y>_sm
- [ ] build_tensorflow_inference_<X.Y>_ec2
- [ ] build_tensorflow_inference_<X.Y>_graviton
</details>

### Additional context

Adding signal handler in pid - 1 to reap zombie process.
Before Fix
```
root@d1aa1e4b8baf:/# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  8.4  3.0 2175388 246528 ?      Ssl  21:52   0:03 python /usr/local/bin/dockerd-entrypoint.py serve.   <--- # inference-toolkit server launcher
root          19  0.1  0.0      0     0 ?        Z    21:52   0:00 [torchserve] <defunct>             <--- # Zombie process when launch frontend
root          20  1.0  0.0      0     0 ?        Z    21:52   0:00 [python] <defunct>                 <--- # Zombie process when running launcher (inference-toolkit)
root          21  7.3  1.7 5605804 138492 ?      Sl   21:52   0:02 java -Dmodel_server_home=/opt/conda/lib/python3.11/site-packages -Dlog4j.configurationFile=file:///opt/conda/lib/python3.11/site-packages/sagemaker_pytorch_serving_container/etc/log4j2.xml -Djava.io.tmpdir=/tmp -XX:-UseContainerSupport -cp .:/opt/conda/lib/python3.11/site-packages/ts.                                   <--- # Frontend Java process
root          57  0.0  0.0   6268  3584 pts/0    Ss   21:52   0:00 /bin/bash               <--- # docker exec -it /bin/bash
root          66  0.0  0.0   7692  3200 pts/0    R+   21:53   0:00 ps aux

after load model
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.6  3.0 2175388 246528 ?      Ssl  21:52   0:07 python /usr/local/bin/dockerd-entrypoint.py serve
root          19  0.0  0.0      0     0 ?        Z    21:52   0:00 [torchserve] <defunct>
root          20  0.0  0.0      0     0 ?        Z    21:52   0:00 [python] <defunct>
root          21  0.6  2.0 6437348 166100 ?      Sl   21:52   0:07 java -Dmodel_server_home=/opt/conda/lib/python3.11/site-packages -Dlog4j.configurationFile=file:///opt/conda/lib/python3.11/site-packages/sagemaker_pytorch_serving_container/etc/log4j2.xml -Djava.io.tmpdir=/tmp -XX:-UseContainerSupport -cp .:/opt/conda/lib/python3.11/site-packages/ts
root          57  0.0  0.0   6268  3584 pts/0    Ss   21:52   0:00 /bin/bash
root         116  8.5  4.2 2335412 344188 ?      Sl   22:11   0:04 /opt/conda/bin/python3.11 /opt/conda/lib/python3.11/site-packages/ts/model_service_worker.py --sock-type unix --sock-name /tmp/.ts.sock.9000 --metrics-config /opt/conda/lib/python3.11/site-packages/ts/configs/metrics.yaml
root         132  0.8  0.0      0     0 ?        Z    22:11   0:00 [python] <defunct>    <---- additional Zombie
root         137  0.0  0.0   7692  3200 pts/0    R+   22:12   0:00 ps aux
```

After Fix
```
root@e0b0dfd82f50:/# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.9  3.0 2175404 246684 ?      Ssl  23:31   0:04 python /usr/local/bin/dockerd-entrypoint.py serve
root          22  0.9  2.4 6585976 194368 ?      Sl   23:31   0:05 java -Dmodel_server_home=/opt/conda/lib/python3.11/site-packages -Dlog4j.configurationFile=file:///opt/conda/lib/python3.11/site-packages/sagemaker_pytorch_serving_container/etc/log4j2.xml -Djava.io.tmpdir=/tmp -XX:-UseContainerSupport -cp .:/opt/conda/lib/python3.11/site-packages/ts
root          68  0.0  0.0   6268  3840 pts/0    Ss   23:32   0:00 /bin/bash
root         131  0.0  0.0   7692  3200 pts/0    R+   23:40   0:00 ps aux

after load model
root@e0b0dfd82f50:/# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.9  3.0 2175404 246684 ?      Ssl  23:31   0:04 python /usr/local/bin/dockerd-entrypoint.py serve
root          22  0.9  3.0 6686344 244952 ?      Sl   23:31   0:05 java -Dmodel_server_home=/opt/conda/lib/python3.11/site-packages -Dlog4j.configurationFile=file:///opt/conda/lib/python3.11/site-packages/sagemaker_pytorch_serving_container/etc/log4j2.xml -Djava.io.tmpdir=/tmp -XX:-UseContainerSupport -cp .:/opt/conda/lib/python3.11/site-packages/ts
root          68  0.0  0.0   6268  3840 pts/0    Ss   23:32   0:00 /bin/bash
root         134 71.2  4.2 2334392 343760 ?      Sl   23:40   0:03 /opt/conda/bin/python3.11 /opt/conda/lib/python3.11/site-packages/ts/model_service_worker.py --sock-type unix --sock-name /tmp/.ts.sock.9001 --metrics-config /opt/conda/lib/python3.11/site-packages/ts/configs/metrics.yaml
root         155  0.0  0.0   7692  3200 pts/0    R+   23:40   0:00 ps aux
```

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py [hunk 1]
```diff
@@ -15,8 +15,29 @@
 import shlex
 import subprocess
 import sys
+import os
+import signal
 
 
+def reap_zombies(signum, frame):
```
Comment (arjkesh) [inline pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py:22]: Could you run a commit to make sure this doesn't break basic functionality?

You can use our command line utility to run tests for a particular image

Steps:
1. Set `autopatch_build` to "False" in https://github.com/aws/deep-learning-containers/blob/master/pytorch/inference/buildspec-2-3-sm.yml#L8
2. Run the command below (assuming your remote is called 'origin')
```
python src/prepare_dlc_dev_environment.py -b pytorch/inference/buildspec-2-3-sm.yml -cp origin 
```
3. When you've confirmed tests run successfully, revert the temp changes with the command below
```
python src/prepare_dlc_dev_environment.py -rb pytorch/inference/buildspec-2-3-sm.yml -cp origin 
```

File: pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py [hunk 2]
```diff
@@ -15,8 +15,29 @@
 import shlex
```
Comment (arjkesh) [inline pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py:15]: @chen3933 does this also need to be added to the ec2 torchserve entrypoint? https://github.com/aws/deep-learning-containers/blob/master/pytorch/inference/docker/build_artifacts/torchserve-ec2-entrypoint.py

Comment (chen3933) [reply pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py:15]: On EC2 customer can run 
```
docker run --init <image_name>
```
which will enable pid 1 to reap zombie child processes.

File: pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py (hunk 1 above)
Comment (chen3933) [reply pytorch/inference/docker/build_artifacts/torchserve-entrypoint.py:22]: The test passed by failed to download report
```

================== 4 passed, 14 warnings in 233.86s (0:03:53) ==================
--
3799 | Downloading Test reports for image: 669063966089.dkr.ecr.us-west-2.amazonaws.com/pr-pytorch-inference:2.3.0-cpu-py311-ubuntu20.04-sagemaker-pr-4249-autopatch-2024-09-18-16-46-42
3800 | Downloading executions cache from ec2 instance
3801 | Downloading executions cache from ec2 instance
3802 | Cache file wasn't downloaded: [Errno 2] No such file
3803 | Cache file wasn't downloaded: [Errno 2] No such file
3804 | No cache file was created
3805 | No cache file was created
3806 | Terminating Instances for image: 669063966089.dkr.ecr.us-west-2.amazonaws.com/pr-pytorch-inference:2.3.0-cpu-py311-ubuntu20.04-sagemaker-pr-4249-autopatch-2024-09-18-16-46-42
3807 | Destroying ssh Key_pair for image: 669063966089.dkr.ecr.us-west-2.amazonaws.com/pr-pytorch-inference:2.3.0-cpu-py311-ubuntu20.04-sagemaker-pr-4249-autopatch-2024-09-18-16-46-42
3808 | Traceback (most recent call last):
3809 | File "test/testrunner.py", line 569, in <module>
3810 | main()
3811 | File "test/testrunner.py", line 557, in main
3812 | run_sagemaker_local_tests(testing_image_list, pytest_cache_params)
3813 | File "test/testrunner.py", line 74, in run_sagemaker_local_tests
3814 | raise RuntimeError(
3815 | RuntimeError: SageMaker Local tests failed on the following DLCs:
3816 | [
3817 | "669063966089.dkr.ecr.us-west-2.amazonaws.com/pr-pytorch-inference:2.3.0-gpu-py311-cu121-ubuntu20.04-sagemaker-pr-4249-autopatch-2024-09-18-16-46-42"
3818 | ]
3819 |  
3820 | [Container] 2024/09/18 17:19:21.326559 Command did not exit successfully python test/testrunner.py exit status 1
3821 | [Container] 2024/09/18 17:19:21.330468 Phase complete: BUILD State: FAILED


```

**Review by arjkesh — APPROVED** [review]

**Pull request merged**